### PR TITLE
Install legendary prerequisites

### PR DESCRIPTION
**Pull request by Etaash-mathamsetty**
Installs the prerequisites that legendary gives in the json when installing a game, we can rip out the dedicated ubisoft stuff with this change. 

Use the following Checklist if you have changed something on the Backend or Frontend:

- [x] Tested the feature and it's working on a current and clean install.
- [x] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

### CODE REVIEW

**Comment by Etaash-mathamsetty** [comment]
the only issue is to move all the setup functions so that they only happen post install or on first run, otherwise the setup functions will constantly run each launch

File: src/backend/main.ts [hunk 1]
```diff
@@ -1099,12 +1093,6 @@ ipcMain.handle(
       return false
     })
 
-    // Stop display sleep blocker
-    if (powerDisplayId !== null) {
-      logInfo('Stopping Display Power Saver Blocker', LogPrefix.Backend)
-      powerSaveBlocker.stop(powerDisplayId)
-    }
```
Comment (arielj) [inline src/backend/main.ts:1106]: is this intentional? looks unrelated to installing pre-requisites

File: src/backend/storeManagers/legendary/setup.ts [hunk 2]
```diff
@@ -30,107 +27,27 @@ export const legendarySetup = async (appName: string) => {
...
-    const oneWeekAgo = new Date()
-    oneWeekAgo.setDate(oneWeekAgo.getDate() - 7)
-    if (cachedAt > oneWeekAgo) {
-      return true
+  if (
+    gameInfo.install?.platform === 'Windows' ||
+    gameInfo.install?.platform === 'Win32' ||
+    gameInfo.install?.platform === 'windows'
+  ) {
+    const info = await getInstallInfo(appName, gameInfo.install.platform)
```
Comment (arielj) [inline src/backend/storeManagers/legendary/setup.ts:35]: this call to getInstallInfo can throw an exception, in other places in the code we do a try/catch just in case

File: src/backend/storeManagers/legendary/setup.ts [hunk 3]
```diff
@@ -30,107 +27,27 @@ export const legendarySetup = async (appName: string) => {
...
+    if (
+      info.manifest.prerequisites &&
```
Comment (arielj) [inline src/backend/storeManagers/legendary/setup.ts:37]: just curious, I see we have this typed as an object, I wonder if there's no game that has this as an array of prerequisites maybe?

the ones I tested they have either 1 prerequisite or null, but I'm curious if we should handle that possibility

**Review by arielj — COMMENTED** [review]
Looks good! I tested with a few games with and without prerequisites (including ubisoft) and all works fine

I just left some comments and one question

**Comment by arielj** [comment]
Testing with RollerCoaster Tycoon 3 I get this error:
image

maybe we need to skip the args for the command when it's an empty string?

```
			"prerequisites": {
				"ids": [
					"DirectXJun2010Redist"
				],
				"name": "DirectX June 2010 Redistributable",
				"path": "directx_Jun2010_redist/DXSETUP.exe",
				"args": ""
			},
```

File: src/backend/storeManagers/legendary/setup.ts [hunk 4]
```diff
@@ -30,107 +27,27 @@ export const legendarySetup = async (appName: string) => {
...
+      info.manifest.prerequisites.path.length > 0
+    ) {
+      await runWineCommandOnGame(appName, {
+        commandParts: [
+          join(
+            gameInfo.install.install_path ?? '',
+            info.manifest.prerequisites.path
+          ),
+          info.manifest.prerequisites.args
+        ],
```
Comment (imLinguin) [inline src/backend/storeManagers/legendary/setup.ts:47]: args is a string right? you should shlex.split them

File: src/backend/storeManagers/legendary/setup.ts (hunk 2 above)
Comment (Etaash-mathamsetty) [reply src/backend/storeManagers/legendary/setup.ts:35]: where? I don't see a try catch anywhere

Comment (Etaash-mathamsetty) [reply src/backend/storeManagers/legendary/setup.ts:36]: there's a try catch within the function itself it looks like

**Comment by Etaash-mathamsetty** [comment]
> Testing with RollerCoaster Tycoon 3 I get this error: image
> 
> maybe we need to skip the args for the command when it's an empty string?
> 
> ```
> 			"prerequisites": {
> 				"ids": [
> 					"DirectXJun2010Redist"
> 				],
> 				"name": "DirectX June 2010 Redistributable",
> 				"path": "directx_Jun2010_redist/DXSETUP.exe",
> 				"args": ""
> 			},
> ```

i think the shlex.split should fix this, could you retest?

File: src/backend/storeManagers/legendary/setup.ts (hunk 2 above)
Comment (arielj) [reply src/backend/storeManagers/legendary/setup.ts:36]: The function has a try/catch inside, but when it catches an error it throws a new one https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/a460b9721910a7a55ed59ccb15aae9992dad15ec/src/backend/storeManagers/legendary/library.ts#L245

I found one getInstallInfo using the try/catch here https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/a460b9721910a7a55ed59ccb15aae9992dad15ec/src/frontend/screens/Library/components/InstallModal/DownloadDialog/index.tsx#L267 but now I see it's a different `getInstallInfo` function, not that one


I do think we should handle that possible throw though (I think the other places that are not handling it are something that we should fix eventually too)

Comment (Etaash-mathamsetty) [reply src/backend/storeManagers/legendary/setup.ts]: makes sense

**Comment by arielj** [comment]
> i think the shlex.split should fix this, could you retest?

yep! the shlex.split fixes that for rollercoaster tycoon

**Review by arielj — APPROVED** [review]
Looks good to me!

I only tested on linux, I understand this has no impact on windows.

Also, the only game that really needed the pre-requisites for me was XCOM 2 and that now works out of the box with this change.

For the other games I own it looks like the pre-requisites are not really needed for the games to work (I didn't need the pre-requisites for any other game), but I think it's good to install them anyway, doesn't seem to create any issue anyway.

File: src/backend/storeManagers/legendary/setup.ts [hunk 5]
```diff
@@ -30,107 +29,31 @@ export const legendarySetup = async (appName: string) => {
...
-const downloadIfNotCached = async (cachePath: string, url: string) => {
-  if (existsSync(cachePath)) {
-    // if the cached file exist but it was cached more than a week ago, download a new one
-    // an outdated installer can make the installation fail according to some report
-    const cachedAt = statSync(cachePath).mtime
-    const oneWeekAgo = new Date()
-    oneWeekAgo.setDate(oneWeekAgo.getDate() - 7)
-    if (cachedAt > oneWeekAgo) {
-      return true
+  if (
```
Comment (flavioislima) [inline src/backend/storeManagers/legendary/setup.ts:32]: Maybe we could just:
```
const winPlatforms = ['windows', 'Windows', 'Win32']
if (winPlatforms.includes(gameInfo.install?.platform)) {
```

**Review by flavioislima — APPROVED** [review]
left a small comment, looks good otherwise.
